### ISSUE TITLE
Problems when parsing <pre> tags (and more)

### ISSUE THREAD

**Issue by christosmarg** [newsboat/newsboat#1429]
## `newsboat` version:

newsboat 2.21.0
System: FreeBSD 12.2-RELEASE-p2 (amd64)
Compiler: g++ 4.2.1 Compatible FreeBSD Clang 8.0.1 (tags/RELEASE_801/final 366581)
ncurses: ncurses 5.9.20140222 (compiled with 5.9)
libcurl: libcurl/7.74.0 OpenSSL/1.1.1d zlib/1.2.11 nghttp2/1.42.0 (compiled with 7.74.0)
SQLite: 3.34.0 (compiled with 3.34.0)
libxml2: compiled with 2.9.10

## Issue
This issue doesn't seem to be OS or terminal specific, but it's been reproduced on FreeBSD and Arch Linux with `st` and `xterm`.

Although I haven't tested it with many articles, except my own, and it'd be rather weird to be article specific, I've noticed that `newsboat` eats up parts of the text when it's wrapped in a `<pre>` tag.

I've noticed two things:
1. It deletes blank lines inside the `<pre>`
2. It might delete parts of a line, or the whole line even. This seems to be happening with lines containing code.

My suspicion for the latter is that the parser gets confused if there's "actual code" to be parsed.

### Blank line deletion
On `newsboat`

![2021-01-13-234947_1112x224_scrot](https://user-images.githubusercontent.com/54286563/104514621-75fa2200-55e9-11eb-8cc9-9afa90633283.png)

Same part of the same article, but on `Feedbro` (Firefox add-on)

![2021-01-13-235024_1001x293_scrot](https://user-images.githubusercontent.com/54286563/104514708-97f3a480-55e9-11eb-808e-b1bf49561f20.png)

### C code inside `<pre>`

`newsboat`

![2021-01-13-235443_508x177_scrot](https://user-images.githubusercontent.com/54286563/104515176-57e0f180-55ea-11eb-9f9a-3b377f3dcc64.png)

`Feedbro`

![2021-01-13-235632_430x187_scrot](https://user-images.githubusercontent.com/54286563/104515219-67f8d100-55ea-11eb-82fe-0c71553f55b5.png)

### Bonus: Random weird positioning of `<p>` and more line deletion inside `<pre>`

`newsboat`

![2021-01-13-235757_1174x502_scrot](https://user-images.githubusercontent.com/54286563/104515477-cb82fe80-55ea-11eb-8aee-29733ac2528d.png)

`Feedbro`

![2021-01-13-235946_1389x502_scrot](https://user-images.githubusercontent.com/54286563/104515531-dccc0b00-55ea-11eb-8c0c-bf5b5a021b8e.png)

**Comment by Minoru** [comment]
Thanks! Can you provide the actual XML that led to this output, so I can reproduce? Note that you can make Newsboat read local files with file://, e.g. put `file:///home/user/reproducer.xml` into your urls file, and Newsboat will treat it as a feed and parse it just as if it was an http:// URL.

**Comment by christosmarg** [comment]
> Can you provide the actual XML that led to this output, so I can reproduce?

https://christosmarg.xyz/rss.xml

> Note that you can make Newsboat read local files

I've tried it but got the same result.

**Comment by Minoru** [comment]
The problem with C code is that the ampersand is escaped only once. When XML parser processes the document, it turns the `&amp;` into `&`; when HTML renderer sees `&`, it assumes that what follows is a name of an HTML entity, doesn't find it, and freaks out. I will accept any patch that makes the parser more resilient, but I don't consider malformed input to be a bug in Newsboat.

The other two are less obvious, I'll take a look at them tomorrow.

**Comment by Minoru** [comment]
The mystery of blank lines is solved too: Newsboat expects empty line to contain a single character, newline, but the feed has some whitespace before the newline: https://github.com/newsboat/newsboat/blob/c282d28504f47d8b2c92009285f5db2481545fcf/src/htmlrenderer.cpp#L911-L925

That'd be easy enough to fix.

**Comment by Minoru** [comment]
Ah, the bonus (weird `<p>` positioning and missing lines) is caused by the same thing as the C code problem: HTML entities are only escaped once. For example, in the following part of the `<description>` tag:

```
        &lt;pre&gt;
        for (pc = (unsigned short *)buf, pos = 0; pos &lt; len; pos++) {
                switch (src[pos]) {
                        .
                        .
                        .
                        .
                }
        }
        &lt;/pre&gt;
```

The less-than sign in the loop condition is encoded as `&lt;`; XML parser turns it into `<`, which makes HTML renderer look for a tag name which isn't there. Again, the correct encoding would be `&amp;lt;` — XML parser will turn that into `&lt;`, which HTML renderer will process as expected.

The closing angle brackets don't stump the HTML renderer, so `&gt;` is rendered as `>`. But `&lt;` always confuses the poor program :)

So again: I'll accept a patch making the renderer (or the parser) more resilient, but I don't consider malformed input to be a Newsboat bug.

I'll close this issue once I fixed the missing blank lines.

**Comment by Minoru** [comment]
@christosmarg, BTW, consider using CDATA blocks for this — it's a neat way to avoid all this escaping (although CDATA can't be nested, and you still have to look out of `]]>` in your posts).

**Comment by christosmarg** [comment]
I updated my whole RSS feed and blog script to wrap *all* `<pre>`s in a `CDATA` block and it all works fine now!

**Comment by Minoru** [comment]
I was wrong about what code causes this — it's `TextFormatter` which turns an array of rendered strings into a markup for our UI library, not `HtmlRenderer` — but anyway, #1431 fixes this bug. Hopefully it'll pass review and will be merged in a few days, and then released as part of 2.23 in March.
